### PR TITLE
Refactor precedence for whitespace changes

### LINKED ISSUE #10516: Refactor parser to support precedence change
Unify space-precedence grouping with operator-precedence resolution. This is the first step of the changes described in #10366.

### PR DESCRIPTION
**Pull request by kazcw**
### Pull Request Description

Single-phase whitespace-aware precedence resolution.

#### Performance
newplot(4)
Since this is a major refactor of the core of the parser, I benchmarked it; it's about 3% faster.

### Important Notes

- Move operator-identifier recognition to lexer.
- Move compound-token assembly out of precedence resolver

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [ ] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [ ] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [ ] Unit tests have been written where possible.

### CODE REVIEW

File: lib/rust/parser/debug/tests/parse.rs [hunk 1]
```diff
@@ -365,6 +366,21 @@ fn type_operator_methods() {
            (Function (OprApp (Ident Foo) (Ok ".") (Ident #"+"))
                      #((() (Ident self) () ()) (() (Ident b) () ())) () "=" (Ident b))))];
     test(&code.join("\n"), expected);
+    test!("Any.==", (OprApp (Ident Any) (Ok ".") (Ident #"==")));
+    expect_invalid_node("x.-y");
+    expect_invalid_node("x.-1");
+    expect_invalid_node("x.~y");
+    expect_invalid_node("x.~1");
+    expect_invalid_node("x.+y");
+    expect_invalid_node("x.+1");
```
Comment (kazcw) [inline lib/rust/parser/debug/tests/parse.rs]: This is the only changed valid parse: It is no longer allowed to apply the dot operator, unspaced, to an RHS consisting of a unary operator application.

File: lib/rust/parser/debug/tests/metadata/mod.rs [hunk 2]
```diff
@@ -1,3 +1,5 @@
+
+
```
Comment (farmaazon) [inline lib/rust/parser/debug/tests/metadata/mod.rs]: Why these blank lines?

File: lib/rust/parser/src/syntax/treebuilding/mod.rs
Comment (farmaazon) [inline lib/rust/parser/src/syntax/treebuilding/mod.rs:1]: I thought we don't use `mod.rs` (`treebuilding.rs` in parent dir instead)?

**Review by farmaazon — APPROVED** [review]

File: lib/rust/parser/debug/tests/metadata/mod.rs (hunk 2 above)
Comment (kazcw) [reply lib/rust/parser/debug/tests/metadata/mod.rs]: Apparently the enso-formatter added them, and the next time I ran it, it took them away.

File: engine/runtime-integration-tests/src/test/java/org/enso/compiler/ErrorCompilerTest.java [hunk 3]
```diff
@@ -52,12 +52,7 @@ public void unaryMinus() throws Exception {
     main = Date.new day=-
     """);
 
-    assertSingleSyntaxError(
-        ir,
-        new Syntax.UnsupportedSyntax("Strange unary -"),
-        "Syntax is not supported yet: Strange unary -",
-        51,
-        52);
+    assertSingleSyntaxError(ir, Syntax.UnrecognizedToken$.MODULE$, "Unrecognized token", 51, 52);
```
Comment (kazcw) [inline engine/runtime-integration-tests/src/test/java/org/enso/compiler/ErrorCompilerTest.java:55]: The error in this case has changed to a multiple-operator error. I think this makes more sense than the previous error type.

**Review by JaroslavTulach — APPROVED** [review]
The changes in tests I see look very nice. Few incompatibilities, but overall most of the  tests continue to work. I see no problem getting this in.

**Pull request merged**